source/simulator/helper_functions.cc

```diff
@@ -2403,6 +2404,235 @@ namespace aspect
+    // Used for constructing the known side of the CBF system.
+    std::vector<double> phi_eta (dofs_per_cell);
+    std::vector<double> phi_rho (dofs_per_cell);
+
+    // Initializing local rhs and mass matrix
+    Vector<double> local_rhs(dofs_per_cell);
+    FullMatrix<double> local_mass_matrix(dofs_per_cell, dofs_per_cell);
+
+    // Loop over all of the surface cells and if one less than h/3 away from
+    // one of the top or bottom boundaries, assemble CBF system for it.
+    typename DoFHandler<dim>::active_cell_iterator
+    cell = dof_handler.begin_active(),
+    endc = dof_handler.end();
+
+    std::vector<types::global_dof_index> local_dof_indices (dofs_per_cell);
+
+    // Get a pointer to the dynamic topography postprocessor.
+    const Postprocess::DynamicTopography<dim> &dynamic_topography =
+      postprocess_manager.template get_matching_postprocessor<Postprocess::DynamicTopography<dim> >();
+
+    // Get the already-computed dynamic topography solution.
+    const LinearAlgebra::BlockVector topo_vector = dynamic_topography.topography_vector();
+    std::vector<double> topo_values( quadrature_formula.size() );
+
+    // TODO: get this from dynamic topography postprocessor
+    const double density_above = 0;
+
+    for (; cell!=endc; ++cell)
+      if (cell->is_locally_owned())
+        {
+          fe_values.reinit (cell);
+          fe_values_adjoint.reinit (cell);
+
+          // Evaluate the material properties and the solution within the cell
+          MaterialModel::MaterialModelInputs<dim> in(fe_values, cell, introspection, solution);
+          MaterialModel::MaterialModelOutputs<dim> out(fe_values.n_quadrature_points, introspection.n_compositional_fields);
+          material_model->evaluate(in, out);
+
+          MaterialModel::MaterialModelInputs<dim> in_adjoint(fe_values, cell, introspection, current_adjoint_solution);
+
+          local_rhs = 0.;
+          local_mass_matrix = 0.;
+
+
+          // VOLUME TERM
+
+          for (unsigned int q=0; q<n_q_points; ++q)
+            {
+              // Get parameters for the density kernel
+              const Tensor<1,dim> velocity_adjoint = in_adjoint.velocity[q];
+              const Tensor<1,dim> gravity = gravity_model->gravity_vector(in.position[q]);
+
+              // Get parameters for the viscosity kernel
+              const double eta = out.viscosities[q];
+              const SymmetricTensor<2,dim> strain_rate_forward = in.strain_rate[q] - 1./3 * trace(in.strain_rate[q]) * unit_symmetric_tensor<dim>();
+              const SymmetricTensor<2,dim> strain_rate_adjoint = in_adjoint.strain_rate[q] - 1./3 * trace(in_adjoint.strain_rate[q]) * unit_symmetric_tensor<dim>();
+
+              for (unsigned int k=0; k<dofs_per_cell; ++k)
+                {
+                  phi_rho[k] = fe_values[introspection.extractors.compositional_fields[0]].value(k,q);
+                  phi_eta[k] = fe_values[introspection.extractors.compositional_fields[1]].value(k,q);
```

Comment:
Here you also want to replace 0 and 1 by the correct compositional indices (compute and store them at the beginning of this function to avoid looking them up for every quadrature point and dof).